File: tests/java/TeamsPayloadTest.java
// test_update_notification.java

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

import org.junit.jupiter.api.Test;

class TeamsNotifierTest{

  record DocumentUpdatedEvent(
      String documentId,
      String documentName,
      String updatedBy,
      String updatedAtIso,
      String channelId
  ) {}

  interface TeamsClient{
    void postMessage(String channelId, String message);
  }

  // unit being tested
  static class notifyTeams{
    static void notifyTeams(DocumentUpdatedEvent event, TeamsClient teamsClient) {
      String message =
          "Document updated: " + event.documentName() + "\n" +
          "Updated by: " + event.updatedBy() + "\n" +
          "Time: " + event.updatedAtIso();

      teamsClient.postMessage(event.channelId(), message);
    }
  }

  @Test
  void testNotificationOfUpdate(){
    var event = DocumentUpdatedEvent(
        "doc-123", 
        "Test Document.xlsx",
        "John Smith",
        "2026-01-29T14:56:00Z",
        "channel-abc"
    );

    TeamsClient teamsClient = mock(TeamsClient.class);

    TeamsNotifier.notifyTeams(event, teamsClient);

    verify(teamsClient, times(1)).postMessage(
        eq("channel-abc"),
        argThat(msg -> msg.contains("Training Schedule.xlsx") && msg.contains("Allison Helling"))
    );
    verifyNoMoreInteractions(teamsClient);
  }
}

File: tests/typescript/teamsPayload.test.ts
import { describe, expect, it, vi } from "vitest";

interface DocumentUpdatedEvent {
  documentId: string;
  documentName: string;
  updatedBy: string;
  updatedAtIso: string;
  channelId: string;
}

interface TeamsClient {
  postMessage(params: { channelId: string; message: string }): void;
}

function notifyTeams(event: DocumentUpdatedEvent, teamsClient: TeamsClient): void {
  const message = [
    `Document updated: ${event.documentName}`,
    `Updated by: ${event.updatedBy}`,
    `Time: ${event.updatedAtIso}`,
  ].join("\n");

  teamsClient.postMessage({ channelId: event.channelId, message });
}

describe("notifyTeams", () => {
  it("sends notification with correct channel and message content", () => {
    const event: DocumentUpdatedEvent = {
      documentId: "doc-123",
      documentName: "Test Document.xlsx",
      updatedBy: "John Smith",
      updatedAtIso: "2026-01-29T14:56:00Z",
      channelId: "channel-abc",
    };

    const teamsClient: TeamsClient = {
      postMessage: vi.fn(),
    };

    notifyTeams(event, teamsClient);

    expect(teamsClient.postMessage).toHaveBeenCalledOnce();
    expect(teamsClient.postMessage).toHaveBeenCalledWith(
      expect.objectContaining({
        channelId: "channel-abc",
        message: expect.stringContaining("Test Document.xlsx"),
      })
    );

    const call = vi.mocked(teamsClient.postMessage).mock.calls[0][0];
    expect(call.message).toContain("John Smith");
  });
});
